ls1intum/artemis-helm
# `Feature`: Profile deployment

Artemis Multiple Instances - Decoupling
========================

## Prerequisites
1. Install kubectl (command line tool to manage Kubernetes clusters)
    - [Windows](https://kubernetes.io/docs/tasks/tools/install-kubectl-windows/)
    - [Mac](https://kubernetes.io/docs/tasks/tools/install-kubectl-macos/)
    - [Linux](https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/)
2. [Install Helm](https://helm.sh/docs/intro/install/) (Kubernetes package manager)
3. [Install minikube](https://minikube.sigs.k8s.io/docs/start/) (Local Kubernetes cluster)


## Setup
- Run `minikube start --cpus 4 --memory 16384` to start the local Kubernetes cluster
- Check that you can connect to the local cluster: `kubectl get nodes`
    - It should show one node
- Enable the Kubernetes ingress within minikube: `minikube addons enable ingress`
- Clone the Artemis Helm Repository: `git clone https://github.com/ls1intum/artemis-helm`
- Checkout the correct branch: `cd artemis-helm && git checkout profile-deployment`
- Create a config file that contains the Helm chart configuration and name it `values.yml` (place it in the top level of the git repository)
```
artemisVersion: "pr-5790"
application:
  userManagement:
    internalAdmin:
      username: "artemis_admin"
      password: "artemis_admin"

  registry:
    jwt: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c"
    password: "SuperSecure"

  # ActiveMQ message broker configuration 
  broker:
    username: "brokeruser"
    password: "brokerpassword"

artemis:
  replicaCount: 0

  image:
    repository: "ghcr.io/ls1intum/"
    name: "artemis"
    pullPolicy: Always

  profileDeployments:
    - profile: "lecture"
      apiPaths:
      - /api/lectures
      - /api/courses/(\d*)/lectures

  ingress:
    className: nginx
    annotations:
      nginx.ingress.kubernetes.io/use-regex: "true"
    hosts:
      - host: artemis.local
        paths:
        - path: "/"
          pathType: Prefix

registry:
  replicaCount: 1
```
- Install the Artemis helm chart: `helm upgrade --install artemis artemis --create-namespace --namespace artemis -f values.yml` (execute the command in the top level of the git repository)
- Get the ip of the ingress: `kubectl get ingress --namespace artemis` (should look something like this)
```
NAME      CLASS   HOSTS           ADDRESS        PORTS   AGE
artemis   nginx   artemis.local   192.168.49.2   80      26m
```
- Create a new line in your `etc/hosts` file: `IP artemis.local` (e.g. `192.168.49.2 artemis-local`)
    - This should work on linux and macos
    - For Windows, you have to edit this file `%windir%\system32\drivers\etc\HOSTS` (the syntax should be identical)

- You should now be able to reach artemis at [http://artemis.local](http://artemis.local)


## Testing Steps
- Log in using the username `artemis_admin` and password `artemis_admin`
- Create a new course
- Verify that the lectures-functionality is not available (e.g. no buttons to create lectures)
![image](https://github.com/ls1intum/artemis-helm/assets/5084100/3e6dcc00-e582-4e07-8fa2-1c6216d6ee0d)
- Start a second instance of artemis with lectures enabled: `kubectl scale deployment artemis-lecture --replicas 1 --namespace artemis`
- After the second instance has started (might take a few minutes depending on your system), the lecture functionality should be available. Try using it
![image](https://github.com/ls1intum/artemis-helm/assets/5084100/c0717b2a-b4b8-40a1-a47e-298395876a59)
- Shut down the second instance: `kubectl scale deployment artemis-lecture --replicas 0 --namespace artemis`. The lecture functionality should be removed again
